docs/proposals/202001_thanos_query_health_handling.md

```diff
@@ -0,0 +1,86 @@
+
+This is problematic in the cache of end-response caching. If we know that certain StoreAPI nodes should always be up then we always want to have not just an error (if partial response is disabled) but also the appropriate `Cache-Control` header in the response. But right now we would only get it for maximum 5 seconds after a store node would go down.
+
+Thus, this logic needs to be changed somehow. There are a few possible options:
+
+* `--store.unhealthy-timeout` could be made to apply to this case as well - we could still consider it a part of the active store set while it is still visible in the UI
+* Another option could be introduced such as `--store.hold-timeout` which would be `--store.unhealthy-timeout`'s brother and we would hold the StoreAPI nodes for `max(hold_timeout, unhealthy_timeout)`.
+* Another option such as `--store.strict-mode` could be introduced which means that we would always retain the last information of the StoreAPI nodes after the first successful check
+* The StoreAPI node specification format that is used in `--store` could be extended to include another flag which would let specify the previous option per-specific node
+
+Lets look through their pros and cons:
+
+* In general it would be nice to avoid adding new options so the first option seems the most attractive in this regard but it is also the most invasive one;
+* Second option increases the configuration complexity the most since you would have to think about the other option `--store.unhealthy-timeout` as well while setting it;
+* The last option is the most complex from code's perspective but it is also least invasive; however it has some downsides like the syntax for specifying store nodes becomes really ugly and hard to understand because we would have not only the DNS query type in there but also a special marker to enable that mode
+
+If we were to graph these choices in terms of theirs incisiveness and complexity it would look something like this:
+
+```text
+Most incisive / Least Complex ------------ Least incisive / Most Complex
+#1           #2                                      #4
+           #3
+```
+
+At first there was an initial attempt to add the last option but it was met with some discussion if this is really the way to go forward. After careful consideration and with the rationale in this proposal we could go with the most invasive choice. But it also probably makes the most sense since it would logically follow that as long as a node is visible in the UI then we send a query to it.
+
+## Goals
+
+* Update the health check logic to properly handle the case when a node disappears in terms of a caching layer
+
+## No Goals
+
+* Fixing the cache handling in the cases where a **new** StoreAPI gets added to the active store set
+
+## Verification
+
+* Unit tests which would fire up a dummy StoreAPI and check different scenarios
+* Ad-hoc testing
+
+## Proposal
```

Comment:
Can we try to think about some solution (or future solution) for discovered notes as @brancz suggested?

Reply:
Or at least let's mention we mean static nodes here and move SD nodes to `Future Work`

Reply:
That's what the blurb in the "Future Work" section at the end is about. Do you think we should mention something more? I have added there some more text that there we are talking about the cases where SD is used.

Reply:
Where? I think we need to explain it better what are the challenges with service discovery.

Reply:
I feel like there was some miscommunication. It was only briefly mentioned in the end before this. I really wanted to keep it brief since all of this will probably come in a separate proposal. Now I have added 4 paragraphs about this under "No Goals". I really think that we shouldn't elaborate more on that in this document since we are not trying to solve that problem now.